File: knapsack-problem/knapsack_problem.clj
(defn knapsack [weights values capacity]
  (let [n (count weights)
        dp (vec (for [i (range (inc n))]
                 (vec (repeat (inc capacity) 0))))]
    (reduce (fn [dp i]
             (reduce (fn [dp w]
                      (let [weight (nth weights (dec i))
                            value (nth values (dec i))]
                        (if (<= weight w)
                          (assoc-in dp [i w] 
                                   (max (get-in dp [(dec i) w])
                                        (+ (get-in dp [(dec i) (- w weight)]) value)))
                          (assoc-in dp [i w] (get-in dp [(dec i) w])))))
                    dp (range 1 (inc capacity))))
           dp (range 1 (inc n)))
    (get-in dp [n capacity])))

(println (knapsack [2 1 3] [4 2 3] 4))

File: knapsack-problem/knapsack_problem.rs
fn knapsack(weights: &[i32], values: &[i32], capacity: i32) -> i32 {
    let n = weights.len();
    let mut dp = vec![vec![0; (capacity + 1) as usize]; n + 1];
    
    for i in 1..=n {
        for w in 1..=capacity {
            if weights[i-1] <= w {
                dp[i][w as usize] = dp[i-1][w as usize].max(
                    dp[i-1][(w - weights[i-1]) as usize] + values[i-1]
                );
            } else {
                dp[i][w as usize] = dp[i-1][w as usize];
            }
        }
    }
    
    dp[n][capacity as usize]
}

fn main() {
    println!("{}", knapsack(&[2, 1, 3], &[4, 2, 3], 4));
}